refactor: Make CWalletTx sync state type-safe

ryanofsky opened the pull request:
Current `CWalletTx` state representation makes it possible to set inconsistent states that won't be handled correctly by wallet sync code or serialized & deserialized back into the same form.

For example, it is possible to call `setConflicted` without setting a conflicting block hash, or `setConfirmed` with no transaction index. And it's possible update individual `m_confirm` and `fInMempool` data fields without setting an overall consistent state that can be serialized and handled correctly.

Fix this without changing behavior by using `std::variant`, instead of an enum and collection of fields, to represent sync state, so state tracking code is safer and more legible.

This is a first step to fixing state tracking bugs https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking, by adding an extra margin of safety that can prevent new bugs from being introduced as existing bugs are fixed.

ryanofsky committed:
refactor: Make CWalletTx sync state type-safe

Current CWalletTx state representation makes it possible to set
inconsistent states that won't be handled correctly by wallet sync code
or serialized & deserialized back into the same form.

For example, it is possible to call setConflicted without setting a
conflicting block hash, or setConfirmed with no transaction index. And
it's possible update individual m_confirm and fInMempool data fields
without setting an overall consistent state that can be serialized and
handled correctly.

Fix this without changing behavior by using std::variant, instead of an
enum and collection of fields, to represent sync state, so state
tracking code is safer and more legible.

This is a first step to fixing state tracking bugs
https://github.com/bitcoin-core/bitcoin-devwiki/wiki/Wallet-Transaction-Conflict-Tracking,
by adding an extra margin of safety that can prevent new bugs from being
introduced as existing bugs are fixed.

promag commented on src/wallet/transaction.h:
```diff
@@ -140,8 +134,7 @@ class CWalletTx
     mutable CAmount nChangeCached;
 
     CWalletTx(const CWallet* wallet, CTransactionRef arg)
```
7bc16a0b421380a1682ca9d97b7f7fc140fb1c5e

nit, `CWalletTx(const CWallet*, CTransactionRef arg)` to make it clear wallet is unused.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/21206#discussion_r579802970

> nit, `CWalletTx(const CWallet*, CTransactionRef arg)` to make it clear wallet is unused.

Good catch! Removed unused argument in this commit instead of later commit

promag reviewed (commented):
Approach ACK (also base PR).

This makes it a lot easier to reason about transaction state. Another great achievement is dropping wallet dependency from transaction. 👏

ryanofsky reviewed (commented):
Rebased 792d72e2a17fa34e173a82171bdaee3ba88f3e74 -> a4bc501a00293a12f0f6d44d85ffb324823b472d ([`pr/txstate.2`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.2) -> [`pr/txstate.3`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.2-rebase..pr/txstate.3)) on top of #21207 pr/txmove.2
Rebased a4bc501a00293a12f0f6d44d85ffb324823b472d -> 56aeb9ff7d5942fdb010db3773e26233e0d1612d ([`pr/txstate.3`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.3) -> [`pr/txstate.4`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.3-rebase..pr/txstate.4)) on top of #21207 pr/txmove.3
Rebased 56aeb9ff7d5942fdb010db3773e26233e0d1612d -> 4c5f91eaf834e3eacc480c8d277e64aebc5ebd4e ([`pr/txstate.4`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.4) -> [`pr/txstate.5`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.4-rebase..pr/txstate.5)) due to conflicts with #21331 and #18842
Rebased 4c5f91eaf834e3eacc480c8d277e64aebc5ebd4e -> ed04456421d8092457fbefd70b77c13a53a3d20f ([`pr/txstate.5`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.5) -> [`pr/txstate.6`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.5-rebase..pr/txstate.6)) on top of #21207 pr/txmove.4 due to conflict with #21141
Rebased ed04456421d8092457fbefd70b77c13a53a3d20f -> 045e712a489ed8df4cc15f29a059c8f661759e72 ([`pr/txstate.6`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.6) -> [`pr/txstate.7`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.6-rebase..pr/txstate.7)) on top of #21207 pr/txmove.5
Rebased 045e712a489ed8df4cc15f29a059c8f661759e72 -> 4e11f88320b644b67db55fe737815451ca7d0681 ([`pr/txstate.7`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.7) -> [`pr/txstate.8`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.7-rebase..pr/txstate.8)) due to conflict with #21666

promag commented on src/wallet/wallet.cpp:
```diff
@@ -963,30 +960,29 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const CWalletTx::Confirmatio
-            // serialized transaction.
-            wtx.m_confirm.block_height = height;
-        } else if (wtx.isConflicted() || wtx.isConfirmed()) {
+        auto lookup_block = [&](const uint256& hash, int& height, TxState& state) {
```
4e11f88320b644b67db55fe737815451ca7d0681

nit, `lookup_block` seems misleading, maybe `update_tx_state` or `mark_inactive`.

promag commented on src/util/overloaded.h:
```diff
@@ -0,0 +1,21 @@
+#include <utility>
+
+//! Overloaded helper for std::visit, useful to write code that switches on a
+//! variant type and triggers a compile error if there are any unhandled cases.
```
4e11f88320b644b67db55fe737815451ca7d0681

I've tested this removing one case in `TxStateSerializedBlockHash` and it failed to compile.

MarcoFalke replied:
Just for reference, our existing use of `std::visit` will also throw compile errors even without this helper. I guess the goal of the helper is to having to write less code in the switch itself?

ryanofsky replied:
> Just for reference, our existing use of `std::visit` will also throw compile errors even without this helper. I guess the goal of the helper is to having to write less code in the switch itself?

I think all existing uses of uses of `std::visit` use custom visitor classes which don't have access to local variables or anything. Or at least they did at the time this was written which was shortly after we enabled c++17. If it makes sense to define visitor classes for a particular variant type, this isn't really meant to be an alternative to that. This is just meant to be an alternative to using `if`/`else if` statements or a `switch` statement. I can update the comment to make it more clear, unless you're suggesting there might be a better way to use `std::visit` without this helper. In that case we could get rid of this.

MarcoFalke replied:
Not saying that this way is better, just saying that it will also emit compile errors even without the helper:

```diff
diff --git a/src/wallet/transaction.h b/src/wallet/transaction.h
index 7dc91bebd8..5a717e498c 100644
--- a/src/wallet/transaction.h
+++ b/src/wallet/transaction.h
@@ -8,12 +8,11 @@
 #include <amount.h>
 #include <primitives/transaction.h>
 #include <serialize.h>
-#include <wallet/ismine.h>
 #include <threadsafety.h>
 #include <tinyformat.h>
-#include <util/overloaded.h>
 #include <util/strencodings.h>
 #include <util/string.h>
+#include <wallet/ismine.h>
 
 #include <list>
 #include <variant>
@@ -85,25 +84,27 @@ static inline TxState TxStateInterpretSerialized(TxStateUnrecognized data)
 //! Get TxState serialized block hash. Inverse of TxStateInterpretSerialized.
 static inline uint256 TxStateSerializedBlockHash(const TxState& state)
 {
-    return std::visit(Overloaded{
-        [](const TxStateConfirmed& confirmed) { return confirmed.confirmed_block_hash; },
-        [](const TxStateConflicted& conflicted) { return conflicted.conflicting_block_hash; },
-        [](const TxStateInMempool& in_mempool) { return uint256::ZERO; },
-        [](const TxStateInactive& inactive) { return inactive.abandoned ? uint256::ONE : uint256::ZERO; },
-        [](const TxStateUnrecognized& unrecognized) { return unrecognized.block_hash; }
-    }, state);
+    struct Overloaded {
+        uint256 operator()(const TxStateConfirmed& confirmed) { return confirmed.confirmed_block_hash; }
+        uint256 operator()(const TxStateConflicted& conflicted) { return conflicted.conflicting_block_hash; }
+        uint256 operator()(const TxStateInMempool& in_mempool) { return uint256::ZERO; }
+        uint256 operator()(const TxStateInactive& inactive) { return inactive.abandoned ? uint256::ONE : uint256::ZERO; }
+        uint256 operator()(const TxStateUnrecognized& unrecognized) { return unrecognized.block_hash; }
+    };
+    return std::visit(Overloaded{}, state);
 }
 
 //! Get TxState serialized block index. Inverse of TxStateInterpretSerialized.
 static inline int TxStateSerializedIndex(const TxState& state)
 {
-    return std::visit(Overloaded{
-        [](const TxStateConfirmed& confirmed) { return confirmed.position_in_block; },
-        [](const TxStateConflicted& conflicted) { return -1; },
-        [](const TxStateInMempool& in_mempool) { return 0; },
-        [](const TxStateInactive& inactive) { return inactive.abandoned ? -1 : 0; },
-        [](const TxStateUnrecognized& unrecognized) { return unrecognized.index; }
-    }, state);
+    struct Overloaded {
+        int operator()(const TxStateConfirmed& confirmed) { return confirmed.position_in_block; }
+        int operator()(const TxStateConflicted& conflicted) { return -1; }
+        int operator()(const TxStateInMempool& in_mempool) { return 0; }
+        int operator()(const TxStateInactive& inactive) { return inactive.abandoned ? -1 : 0; }
+        int operator()(const TxStateUnrecognized& unrecognized) { return unrecognized.index; }
+    };
+    return std::visit(Overloaded{}, state);
 }

promag commented on src/wallet/transaction.h:
```diff
@@ -11,12 +11,99 @@
+struct TxStateInactive {
+    bool abandoned;
+
+    explicit TxStateInactive(bool abandoned=false) : abandoned(abandoned) {}
```
4e11f88320b644b67db55fe737815451ca7d0681

nit, space around `=`.

promag commented on src/wallet/wallet.cpp:
```diff
@@ -963,30 +960,29 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const CWalletTx::Confirmatio
     }
     // If wallet doesn't have a chain (e.g wallet-tool), don't bother to update txn.
     if (HaveChain()) {
         bool active;
```
4e11f88320b644b67db55fe737815451ca7d0681

Make local in lambda below?

promag reviewed (commented):
Code review and tested ACK 4e11f88320b644b67db55fe737815451ca7d0681 (including base PR). Some nits for your consideration.

I think this approach is superior simply because it prevents state inconsistency. Maybe a followup could be to make state members const.

Another nice touch in this PR is `SyncTxState` type which makes it clear which states are handled/accepted in `SyncTransaction`.

ryanofsky reviewed (commented):
Thanks for the review! I really appreciate it on a noisy PR like this where there are so many small changes.

---

Rebased 4e11f88320b644b67db55fe737815451ca7d0681 -> a5a537fd95084a2e2912d9625cef41f5f6732efc ([`pr/txstate.8`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.8) -> [`pr/txstate.9`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.8-rebase..pr/txstate.9)) on top of #21207 pr/txmove.6 due to conflict with #21759, also making some suggested changes

S3RK reviewed (commented):
Not a full review, just a small thing I've noticed looking at #21207

ryanofsky commented on src/wallet/receive.h:
```diff
@@ -10,11 +10,55 @@
+
+bool ScriptIsChange(const CWallet& wallet, const CScript& script) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+bool OutputIsChange(const CWallet& wallet, const CTxOut& txout) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+CAmount OutputGetChange(const CWallet& wallet, const CTxOut& txout) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+CAmount TxGetChange(const CWallet& wallet, const CTransaction& tx);
```
re: https://github.com/bitcoin/bitcoin/pull/21206#discussion_r625576746

> No need to export `OutputGetChange` and `TxGetChange` as they are not used outside of `receive.cpp`

I think in the future it might be good to move low-level code using `IsChange` logic inside `receive.cpp` so `IsChange` functions could be private. But for `GetChange` it seems most consistent to try to keeping exposing `Get{Credit,Debit,Change}` functions as a group. The fact that these two `GetChange` functions don't have callers I think is more a reflection of poor unit test coverage than that change amounts should be a private part of transaction accounting.

ryanofsky reviewed (commented):
Thanks for checking this!

---

Rebased a5a537fd95084a2e2912d9625cef41f5f6732efc -> 3416ade015914ddeb84fad23d2455d974e9f4a74 ([`pr/txstate.9`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.9) -> [`pr/txstate.10`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.9-rebase..pr/txstate.10)) on top of #21207 pr/txmove.7 due to conflicts
Rebased 3416ade015914ddeb84fad23d2455d974e9f4a74 -> 7cc2549759c12182063fce56161ce4403502312e ([`pr/txstate.10`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.10) -> [`pr/txstate.11`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.10-rebase..pr/txstate.11)) on top of #21207 pr/txmove.8 due to conflicts with #17331
Rebased 7cc2549759c12182063fce56161ce4403502312e -> c2b3b4b184148447b695ff6ec56bbc2c24447655 ([`pr/txstate.11`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.11) -> [`pr/txstate.12`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.11-rebase..pr/txstate.12)) on top of #22100 pr/txfun.1 after #21207 merge
Rebased c2b3b4b184148447b695ff6ec56bbc2c24447655 -> c00b540d3e315d52f171678e3e6d5e36211d3c38 ([`pr/txstate.12`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.12) -> [`pr/txstate.13`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.12-rebase..pr/txstate.13)) on top of #22100 pr/txfun.2
Rebased c00b540d3e315d52f171678e3e6d5e36211d3c38 -> b71a25b492d8ded339351ed770a4360ec2316b6c ([`pr/txstate.13`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.13) -> [`pr/txstate.14`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.13-rebase..pr/txstate.14)) on top of #22100 pr/txfun.3 due to conflict with #21866
Rebased b71a25b492d8ded339351ed770a4360ec2316b6c -> 730df28e00822f06c9fc11c86a74f764a2f5136c ([`pr/txstate.14`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.14) -> [`pr/txstate.15`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.14-rebase..pr/txstate.15)) on top of #22100 pr/txfun.4

MarcoFalke commented on src/wallet/wallet.cpp:
```diff
@@ -1187,19 +1181,19 @@ void CWallet::SyncTransaction(const CTransactionRef& ptx, CWalletTx::Confirmatio
     auto it = mapWallet.find(tx->GetHash());
     if (it != mapWallet.end()) {
-        it->second.fInMempool = true;
+        CHECK_NONFATAL(it->second.InMempool());
```
This will fail. See #22359

jonatack commented on src/wallet/transaction.h:
```diff
@@ -11,18 +11,104 @@
+}
+
+//! Get TxState serialized block hash. Inverse of TxStateInterpretSerialized.
+static uint256 TxStateSerializedBlockHash(const TxState& state)
```
730df28e0082 seeing the following warnings when debug-building with debian clang 11

```rake
./wallet/transaction.h:221:59: warning: field 'tx' will be initialized after field 'm_state' [-Wreorder-ctor]
    CWalletTx(CTransactionRef tx, const TxState& state) : tx(std::move(tx)), m_state(state)
                                                          ^
./wallet/transaction.h:71:16: warning: 'static' function 'TxStateInterpretSerialized' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
static TxState TxStateInterpretSerialized(TxStateUnrecognized data)
               ^
./wallet/transaction.h:86:16: warning: 'static' function 'TxStateSerializedBlockHash' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
static uint256 TxStateSerializedBlockHash(const TxState& state)
               ^
./wallet/transaction.h:98:12: warning: 'static' function 'TxStateSerializedIndex' declared in header file should be declared 'static inline' [-Wunneeded-internal-declaration]
static int TxStateSerializedIndex(const TxState& state)
           ^
4 warnings generated.
```

ryanofsky replied:
> [730df28](https://github.com/bitcoin/bitcoin/commit/730df28e00822f06c9fc11c86a74f764a2f5136c) seeing the following warnings when debug-building with debian clang 11

Thanks! Should be fixed in new push

ryanofsky reviewed (commented):
Updated 730df28e00822f06c9fc11c86a74f764a2f5136c -> c5ee3fb0b1db4095376bcff074741dc7a135ba8a ([`pr/txstate.15`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.15) -> [`pr/txstate.16`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.15..pr/txstate.16)) fixing reported compiler warnings and improving std::visit comment
Rebased c5ee3fb0b1db4095376bcff074741dc7a135ba8a -> 8f25d3a6b7a6cc308eb2f0ebc944e1dd425902f5 ([`pr/txstate.16`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.16) -> [`pr/txstate.17`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.16-rebase..pr/txstate.17)) due to conflict with #22492
Rebased 8f25d3a6b7a6cc308eb2f0ebc944e1dd425902f5 -> c5599e514beb28518e4f31f495016cf6d8cd4c24 ([`pr/txstate.17`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.17) -> [`pr/txstate.18`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.17-rebase..pr/txstate.18)) due to conflict with #22359
Rebased c5599e514beb28518e4f31f495016cf6d8cd4c24 -> 095b1d6f58e2302f229617a2439e5a57e14fe936 ([`pr/txstate.18`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.18) -> [`pr/txstate.19`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.18-rebase..pr/txstate.19)) on top of pr/txfun.6 #22100
Rebased 095b1d6f58e2302f229617a2439e5a57e14fe936 -> b17ba1c752a65d2f8b3d1485074d08d48c1ffd3e ([`pr/txstate.19`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.19) -> [`pr/txstate.20`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.20), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.19-rebase..pr/txstate.20)) on top of pr/txfun.7 #22100

promag commented:
Base PR is merged, could update OP and rebase to move this change forward.

ryanofsky commented:
> Base PR is merged, could update OP and rebase to move this change forward.

Thanks, rebased and removed references to base PR in the OP.

Rebased b17ba1c752a65d2f8b3d1485074d08d48c1ffd3e -> 2896ad570523842245d2de77de531776cfcc08d8 ([`pr/txstate.20`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.20) -> [`pr/txstate.21`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.20-rebase..pr/txstate.21)) after base PR #22100 merge

meshcollider reviewed (commented):
Code review ACK 541d5f5c775a9aa884c1a8b37fc123933fdbd98e

Looks great!

ryanofsky commented:
Rebased 3c6fdae6ed67426bd9ddb2949a9e66934f71bf0b -> 78c8ac87e2f585772f296b943c733bd26f255ba0 ([`pr/txstate.24`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.24) -> [`pr/txstate.25`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.24-rebase..pr/txstate.25)) due to conflict with #22928, also making minor `MakeWalletTxStatus` cleanup

meshcollider reviewed (commented):
re-utACK 78c8ac87e2f585772f296b943c733bd26f255ba0

laanwj commented:
I like this refactor, it abstracts the transaction state much better than the boolean-radio-buttons, and agree it makes some kinds of mistake harder to make.

Concept and code review ACK 78c8ac87e2f585772f296b943c733bd26f255ba0

promag reviewed (commented):
Code review ACK 78c8ac87e2f585772f296b943c733bd26f255ba0.

In a follow-up, we can consolidate updates to `m_state`, for instance, the 2nd update can be avoided:
```
transactionAddedToMempool
    SyncTransaction
        AddToWalletIfInvolvingMe
            AddToWallet
                m_state = ...
    RefreshMempoolStatus
        m_state = ...
```

jonatack commented on src/wallet/transaction.h:
```diff
@@ -129,44 +211,12 @@ class CWalletTx
         nTimeSmart = 0;
         fFromMe = false;
         fChangeCached = false;
-        fInMempool = false;
```
There appear to be remaining `fInMempool` comments that may need updating.

```cpp
src/wallet/wallet.cpp:1733:    // We must set fInMempool here - while it will be re-set to true by the
src/wallet/wallet.cpp-1734-    // entered-mempool callback, if we did not there would be a race where a
src/wallet/wallet.cpp-1735-    // user could call sendmoney in a loop and hit spurious out of funds errors
src/wallet/wallet.cpp-1736-    // because we think that this newly generated transaction's change is
src/wallet/wallet.cpp-1737-    // unavailable as we're not yet aware that it is in the mempool.
src/wallet/wallet.cpp-1738-    //
src/wallet/wallet.cpp:1739:    // Irrespective of the failure reason, un-marking fInMempool
src/wallet/wallet.cpp-1740-    // out-of-order is incorrect - it should be unmarked when
src/wallet/wallet.cpp-1741-    // TransactionRemovedFromMempool fires.
--
src/wallet/wallet.cpp:1973    // Get the inserted-CWalletTx from mapWallet so that the
src/wallet/wallet.cpp:1974    // fInMempool flag is cached properly
```

jonatack commented on src/util/overloaded.h:
```diff
@@ -0,0 +1,23 @@
+#define BITCOIN_UTIL_OVERLOADED_H
+
+#include <optional>
+#include <utility>
```
Should these two includes be replaced by `#include <variant>`?

jonatack reviewed (commented):
Approach ACK / light code review ACK 78c8ac87e2f585772f296b943c733bd26f255ba0

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -990,30 +989,29 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const CWalletTx::Confirmatio
-            // serialized transaction.
-            wtx.m_confirm.block_height = height;
-        } else if (wtx.isConflicted() || wtx.isConfirmed()) {
+        auto lookup_block = [&](const uint256& hash, int& height, TxState& state) {
```
```suggestion
        auto lookup_block = [&](const uint256& hash, int height, TxState& state) {
```

jonatack replied:
Oh ok, thanks. I didn't see that `findBlock()` uses `height` as an out-param, but now had a better look at the `FoundBlock` class.

ryanofsky commented on src/wallet/wallet.h:
```diff
@@ -576,7 +576,7 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
 
     /** Pass this transaction to node for mempool insertion and relay to peers if flag set to true */
-    bool SubmitTxMemoryPoolAndRelay(const CWalletTx& wtx, std::string& err_string, bool relay) const;
+    bool SubmitTxMemoryPoolAndRelay(CWalletTx& wtx, std::string& err_string, bool relay) const;
```
re: https://github.com/bitcoin/bitcoin/pull/21206#discussion_r754553318

> Perhaps update the Doxygen documentation in the previous line to explain why `wtx` is after this change mutable / an out-param.

I'll add your text if you want to make a specific suggestion, but mentioning the wtx state update seems like an unusual detail for this summary comment to focus on when it doesn't going into other details like this, and there are already lots of comments about refreshing the in-mempool state. Just changing this to a mutable reference already seems more straightforward than the previous version which used a const reference and mutable state member.

ryanofsky reviewed (commented):
Thanks for reviews and suggestions!

Updated 78c8ac87e2f585772f296b943c733bd26f255ba0 -> d8ee8f3cd32bbfefec931724f5798cbb088ceb6f ([`pr/txstate.25`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.25) -> [`pr/txstate.26`](https://github.com/ryanofsky/bitcoin/commits/pr/txstate.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/txstate.25..pr/txstate.26))

jonatack commented on src/wallet/test/wallet_tests.cpp:
```diff
@@ -367,13 +364,13 @@ static int64_t AddTx(ChainstateManager& chainman, CWallet& wallet, uint32_t lock
-    // is needed before confirm again with different block.
-    return wallet.AddToWallet(MakeTransactionRef(tx), confirm, [&](CWalletTx& wtx, bool /* new_tx */) {
-        wtx.setUnconfirmed();
+    return wallet.AddToWallet(MakeTransactionRef(tx), state, [&](CWalletTx& wtx, bool /* new_tx */) {
```
(if need to retouch)
```suggestion
    return wallet.AddToWallet(MakeTransactionRef(tx), state, [&](CWalletTx& wtx, /*new_tx=*/bool) {
```

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -990,30 +989,29 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const CWalletTx::Confirmatio
+        auto lookup_block = [&](const uint256& hash, int& height, TxState& state) {
             // If tx block (or conflicting block) was reorged out of chain
             // while the wallet was shutdown, change tx status to UNCONFIRMED
             // and reset block height, hash, and index. ABANDONED tx don't have
```
Should the documentation be updated here as well?

jonatack commented on src/wallet/rpcdump.cpp:
```diff
@@ -369,11 +369,9 @@ RPCHelpMan importprunedfunds()
     CTransactionRef tx_ref = MakeTransactionRef(tx);
     if (pwallet->IsMine(*tx_ref)) {
-        pwallet->AddToWallet(std::move(tx_ref), confirm);
+        pwallet->AddToWallet(std::move(tx_ref), TxStateConfirmed{merkleBlock.header.GetHash(), height, (int)txnIndex});
```
Yes, the developer-notes mention to follow the CppCoreGuidelines and there are some advantages to named casts (https://github.com/bitcoin/bitcoin/pull/20965#issuecomment-770375623) but I do hesitate to mention them due to the verbosity (though if casts are seen as a code smell, verbosity could be a good thing, I guess). Thanks!

jonatack reviewed (commented):
Code review ACK d8ee8f3cd32bbfefec931724f5798cbb088ceb6f

Mitumaru reviewed (commented):
You

hebasto commented on src/util/overloaded.h:
```diff
@@ -0,0 +1,22 @@
+//! https://en.cppreference.com/w/cpp/utility/variant/visit#Example
+template<class... Ts> struct Overloaded : Ts... { using Ts::operator()...; };
+
+//! Explicit deduction guide (not needed as of C++20)
```
As for Clang, this explicit deduction guide can be dropped for versions >=17 only, unfortunately.

fanquake replied:
This is part of #29042.
